docs/source/migration.rst

```diff
@@ -9,6 +9,91 @@ to the code affected. Most users should only need to make minimal changes.
 If you've supplied ``PyquilConfig`` objects to functions (or used the ``QVM_URL`` and ``QUILC_URL`` environment variables)
 to override configuration, see :ref:`pyquil_configuration`.
 
-Lastly, pyQuil v3 relies on an updated authentication model. To get going smoothly, you should install the new `QCS CLI
-<https://docs.rigetti.com/en/command-line-interface/command-line-interface>`_ and log in with it before using pyQuil v3
-against real QPUs.
+Authentication
+--------------
+
+pyQuil v3 relies on an updated authentication model. To get started, install the new `QCS CLI
+<https://docs.rigetti.com/>`_  and
+log in with it before using pyQuil v3 with QCS and live QPUs.
+
+
+Parameters & Memory
+-------------------
+
+In order to give the user more control over and visibility into program execution, especially in
+parallel, objects such as ``QuantumComputer``, ``QAM``, ``QPU``, and ``QVM`` are no longer stateful
+with respect to individual programs. These objects are now safe to
+share among different threads, so you can execute and retrieve results in parallel for even better
+performance. (See :doc:`advanced_usage` for more information).
+
+However, this required three small but important changes:
+
+1. ``write_memory`` is no longer a method on ``QAM`` but rather on ``Program`` and ``EncryptedProgram``.
+2. ``qc.run()`` no longer accepts a ``memory_map`` argument. All memory values must be set directly
+  on the ``Program`` or ``EncryptedProgram`` using ``write_memory``.
+3. ``QAM.load()``, ``QAM.wait()``, and ``QAM.reset()`` no longer exist, because the
+  ``QAM`` no longer "stores" program state.
+
+This means that you should now execute your programs using one of these options:
+
+.. code:: python
+
+   qc = get_qc("Aspen-X")
+   program = Program()
+   theta = program.declare('theta', 'REAL')
+   program += RZ(theta, 0)
+   exe = qc.compile(program)
+
+   # Previously, we would have called ``qc.qam.write_memory`` instead
+   exe.write_memory(region_name='theta', value=np.pi)
+
+   # Option 1
+   result = qc.run(exe)
+
+   # Option 2
+   result = qc.qam.run(exe)
+
+   # Option 3
+   job = qc.qam.execute(exe)
```

Comment:
You may want to document `execute()` alongside the existing docs for `run()` (https://github.com/rigetti/pyquil/blob/rc/docs/source/qvm.rst#the-run-method). It's a little weird too that that file is  named `qvm.rst` because it's talking about both qvm and qpu.

Reply:
Good point. added.